### PR TITLE
Refactor predicates in Model & Improve the test structure

### PR DESCRIPTION
**Pull request by UncleGrandpa925**

### CODE REVIEW

File: src/main/java/trackitnus/commons/core/Messages.java [hunk 1]
```diff
@@ -25,4 +23,8 @@
     public static final String MESSAGE_NOT_EDITED = "At least one field to edit must be provided.";
     public static final String MESSAGE_DUPLICATE_MODULE = "This module already exists";
     public static final String MESSAGE_DUPLICATE_TASK = "This task already exists";
+    public static final String MESSAGE_INVALID_INDEX = "Index is not a non-zero unsigned integer.";
+    public static final String DATE_MESSAGE_CONSTRAINTS = "Date should be in the format dd/MM/yyyy or dd/MM/yyyy hh:mm";
```
Comment (simonteozw) [inline src/main/java/trackitnus/commons/core/Messages.java]: I think we should remove the `dd/MM/yyyy hh:mm`. Or I can do that in my own PR later

File: src/main/java/trackitnus/logic/parser/lesson/AddLessonCommandParser.java [hunk 2]
```diff
@@ -43,8 +43,7 @@ public AddLessonCommand parse(String args) throws ParseException {
             ArgumentTokenizer.tokenize(args,
                 PREFIX_CODE, PREFIX_TYPE, PREFIX_DATE);
 
-        if (!arePrefixesPresent(argMultimap, PREFIX_CODE, PREFIX_TYPE, PREFIX_DATE)
-            || !argMultimap.getPreamble().isEmpty()) {
+        if (!arePrefixesPresent(argMultimap, PREFIX_CODE, PREFIX_TYPE, PREFIX_DATE)) {
```
Comment (simonteozw) [inline src/main/java/trackitnus/logic/parser/lesson/AddLessonCommandParser.java:46]: Why did you remove `!argMultimap.getPreamble().isEmpty()`?

File: src/main/java/trackitnus/model/util/SampleDataUtil.java [hunk 3]
```diff
@@ -38,6 +39,8 @@
             new Contact(new Name("Irfan Ibrahim"), new Phone("92492021"), new Email("irfan@example.com"),
                 getTagSet("classmates", "CS2103T")),
             new Contact(new Name("Roy Balakrishnan"), new Phone("92624417"), new Email("royb@example.com"),
+                getTagSet("colleagues")),
+            new Contact(new Name("Henry CHIA Wai Kit"), new Phone("65161556"), new Email("hchia@comp.nus.edu.sg"),
```
Comment (simonteozw) [inline src/main/java/trackitnus/model/util/SampleDataUtil.java:43]: Is the CHIA intended to be all caps? Haha

**Review by simonteozw — COMMENTED** [review]
Looks good! Just some comments

File: src/main/java/trackitnus/model/util/SampleDataUtil.java (hunk 3 above)
Comment (UncleGrandpa925) [reply src/main/java/trackitnus/model/util/SampleDataUtil.java:43]: Yeah I copied his name from SoC's website 😄

File: src/main/java/trackitnus/logic/parser/lesson/AddLessonCommandParser.java (hunk 2 above)
Comment (UncleGrandpa925) [reply src/main/java/trackitnus/logic/parser/lesson/AddLessonCommandParser.java:46]: because argMultimap.getPreamble() is actually to check if index is present, which is unnecessary in this case (in fact this is leftovers of copying from AB3 commands)

File: src/main/java/trackitnus/commons/core/Messages.java (hunk 1 above)
Comment (UncleGrandpa925) [reply src/main/java/trackitnus/commons/core/Messages.java]: Sure!

File: src/main/java/trackitnus/logic/parser/ParserUtil.java [hunk 4]
```diff
@@ -198,9 +199,26 @@ public static LocalDate parseDate(String date) throws ParseException {
...
+     * is valid!). The function is written so that the caller doesn't need to handle the possible exception since
+     * it's expected that no exception will be thrown.
+     */
+    public static LocalDate parseValidDate(String date) {
+        requireNonNull(date);
+        String trimmedDate = date.trim();
+        try {
+            return LocalDate.parse(trimmedDate, DATE_PATTERN);
+        } catch (DateTimeParseException e) {
+            throw new IllegalArgumentException("an invalid date is passed into parseValidDate");
```
Comment (simonteozw) [inline src/main/java/trackitnus/logic/parser/ParserUtil.java]: Hmm Actually how is this different from the normal parseDate? And why not just use the same Messages.DATE_MESSAGE_CONSTRAINTS? In any case, I think the first letter should be capitalised: `An invalid date...`

**Review by simonteozw — COMMENTED** [review]
Looks good! Just 1 comment

File: src/main/java/trackitnus/logic/parser/ParserUtil.java (hunk 4 above)
Comment (UncleGrandpa925) [reply src/main/java/trackitnus/logic/parser/ParserUtil.java]: it is very different because the IllegalArgumentException is an unchecked exception, which doesn't require the caller to try/catch for it. In other words, this function will crash the program (assert-like) if the provided date is invalid!

Comment (UncleGrandpa925) [reply src/main/java/trackitnus/logic/parser/ParserUtil.java]: Also I will capitalize the "a" now

Comment (UncleGrandpa925) [reply src/main/java/trackitnus/logic/parser/ParserUtil.java]: Done!

Comment (simonteozw) [reply src/main/java/trackitnus/logic/parser/ParserUtil.java]: Sorry actually for grammar it should be "An invalid date has been passed into parseValidDate"

Comment (UncleGrandpa925) [reply src/main/java/trackitnus/logic/parser/ParserUtil.java]: done again!

**Review by simonteozw — APPROVED** [review]
LGTM
